pkg/waiter/waiter_test.go

```diff
@@ -0,0 +1,197 @@
+package waiter
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestStringWaiter(t *testing.T) {
+	wm := NewManager[string]()
+	wm.Start(context.Background())
+
+	w, err := wm.NewWaiter()
+	assert.NoError(t, err)
+
+	want := "hello"
+	require.NoError(t, wm.Send(w.ID(), want, nil))
+
+	got, err := w.Wait(context.Background())
+	assert.NoError(t, err)
+	assert.Equal(t, got, want)
+}
+
+func TestStructWaiter(t *testing.T) {
+	type widget struct {
+		msg string
+	}
+
+	wm := NewManager[widget]()
+	wm.Start(context.Background())
+
+	w, err := wm.NewWaiter()
+	assert.NoError(t, err)
+
+	want := widget{msg: "hello"}
+	require.NoError(t, wm.Send(w.ID(), want, nil))
+
+	got, err := w.Wait(context.Background())
+	assert.NoError(t, err)
+	assert.Equal(t, got.msg, want.msg)
+}
+
+func TestPointerWaiter(t *testing.T) {
+	type widget struct {
+		msg string
+	}
+
+	wm := NewManager[*widget]()
+	wm.Start(context.Background())
+
+	w, err := wm.NewWaiter()
+	assert.NoError(t, err)
+
+	want := &widget{msg: "hello"}
+	require.NoError(t, wm.Send(w.ID(), want, nil))
+
+	got, err := w.Wait(context.Background())
+	assert.NoError(t, err)
+	assert.Equal(t, got, want)
+	if got != want {
+		t.Errorf("%p != %p", got, want)
+	}
+}
+
+func TestWaitCancel(t *testing.T) {
+	t.Parallel()
+	wm := NewManager[string]()
+	wm.Start(context.Background())
+
+	w, err := wm.NewWaiter()
+	assert.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		time.Sleep(200 * time.Millisecond)
+		cancel()
+	}()
+
+	got, err := w.Wait(ctx)
+	assert.ErrorContains(t, err, "context canceled")
+	assert.Zero(t, got)
+
+	// allow time for manager to cleanup canceled waiters
+	time.Sleep(100 * time.Millisecond)
+	assert.Equal(t, wm.len(), 0)
+}
+
+func TestWaitClose(t *testing.T) {
+	t.Parallel()
+	wm := NewManager[string]()
+	wm.Start(context.Background())
+
+	w, err := wm.NewWaiter()
+	assert.NoError(t, err)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := w.Wait(context.Background())
+		assert.ErrorIs(t, err, ErrWaiterClosed)
+	}()
+
+	assert.Equal(t, wm.len(), 1)
+	w.Close()
+	wg.Wait()
+
+	// allow time for manager to cleanup closed waiters
+	time.Sleep(200 * time.Millisecond)
+	assert.Equal(t, wm.len(), 0)
+}
+
+func TestCloseManager(t *testing.T) {
+	t.Parallel()
+	wm := NewManager[string]()
+	ctx, cancel := context.WithCancel(context.Background())
+	wm.Start(ctx)
+
+	w, err := wm.NewWaiter()
+	require.NoError(t, err)
+
+	go func() {
+		time.Sleep(200 * time.Millisecond)
+		cancel()
+	}()
+
+	_, err = w.Wait(context.Background())
+	require.ErrorIs(t, err, ErrWaiterClosed)
```

Comment:
and here and below

Reply:
Seems for ErrorIs it's the reverse
![image](https://github.com/stackrox/stackrox/assets/119438707/2137553a-a929-4823-95db-d2b60b6dad39)